examples/ddp2pipe/ddp2pipe.py

```diff
@@ -0,0 +1,299 @@
+        return logits, loss
+
+
+class DDP2PipeConnector(nn.Module):
+    def __init__(self, pp_ranks, pp_group, dp_group_size):
+        super().__init__()
+        self.pp_ranks = pp_ranks
+        self.pp_group = pp_group
+
+        rank = torch.distributed.get_rank()
+        driver_rank = self.pp_ranks[0]
+
+        if rank == driver_rank:
+            self.pipe_model = PipeModel()
+
+            for i in range(1, 4):
+                annotate_split_points(self.pipe_model, {f'mod.{i}': PipeSplitWrapper.SplitPoint.BEGINNING})
+
+            # print(self.pipe_model)
+
+            self.pipe = Pipe.from_tracing(self.pipe_model, output_loss_value_spec=(False, True))
+            args_chunk_spec = (TensorChunkSpec(0), TensorChunkSpec(0))
+            kwargs_chunk_spec = {}
+            output_chunk_spec = (TensorChunkSpec(0), CustomReducer(torch.tensor(0.0), lambda a, b: a + b))
+            all_worker_ranks = self.pp_ranks
+            chunks = len(all_worker_ranks)
+
+            # print(self.pipe)
+
+            self.pipeline = PipelineDriverFillDrain(self.pipe, chunks,
+                                                    args_chunk_spec, kwargs_chunk_spec, output_chunk_spec,
+                                                    world_size=len(all_worker_ranks),
+                                                    all_ranks=all_worker_ranks,
+                                                    _debug_mask_minibatches=False)
+
+            self.pipeline.init_data_parallel(dp_group_size)
+
+            self.optimizer = self.pipeline.instantiate_optimizer(optim.Adam)
+
+    def forward(self, x, labels):
+        rank = torch.distributed.get_rank()
+        driver_rank = self.pp_ranks[0]
+
+        if rank == driver_rank:
+            x_list = [torch.empty_like(x) for _ in range(len(self.pp_ranks))]
+            labels_list = [torch.empty_like(labels) for _ in range(len(self.pp_ranks))]
+        else:
+            x_list = None
+            labels_list = None
+
+        torch.distributed.gather(x, gather_list=x_list, dst=driver_rank, group=self.pp_group)
+        torch.distributed.gather(labels, gather_list=labels_list, dst=driver_rank, group=self.pp_group)
+
+        if rank == driver_rank:
+            x_concat = torch.cat(x_list)
+            x_concat.requires_grad = True
+            labels_concat = torch.cat(labels_list)
+            logits_concat, loss = self.pipeline(x_concat, labels_concat)
+            logits_chunks = list(logits_concat.chunk(len(self.pp_ranks)))
+        else:
+            logits_chunks = None
+            loss = torch.empty(1)
+
+        logits = torch.empty((x.size(0), DIMS[-1]))
+        torch.distributed.scatter(logits, logits_chunks, src=driver_rank, group=self.pp_group)
+        torch.distributed.broadcast(loss, src=driver_rank, group=self.pp_group)
+
+        if torch.is_grad_enabled():
+            if rank == driver_rank:
+                last_grads_chunks = [x[0] for x in self.pipeline.last_grads]  # TODO
```

Comment:
nit: maybe good to write out this TODO?
(ignore if you need to land first)

Reply:
Assuming the TODO here would be to create a torch.function as discussed?